x/subscription/keeper/subscription.go

```diff
@@ -321,6 +333,76 @@ func (k Keeper) advanceMonth(ctx sdk.Context, subkey []byte) {
 	}
 }
 
+func (k Keeper) verifySubExists(ctx sdk.Context, consumer string, block uint64, sub *types.Subscription) bool {
+	if found := k.subsFS.FindEntry(ctx, consumer, block, sub); !found {
+		// subscription (monthly) timer has expired for an unknown subscription:
+		// either the timer was set wrongly, or the subscription was incorrectly
+		// removed; and we cannot even return an error about it.
+		utils.LavaFormatError("critical: month expiry for unknown subscription, skipping",
+			fmt.Errorf("subscription not found"),
+			utils.Attribute{Key: "consumer", Value: consumer},
+			utils.Attribute{Key: "block", Value: block},
+		)
+		return false
+	}
+	return true
+}
+
+func (k Keeper) addCuTrackerTimerForSubscription(ctx sdk.Context, block uint64, sub *types.Subscription) {
+	blocksToSave, err := k.epochstorageKeeper.BlocksToSave(ctx, block)
+	if err != nil {
+		utils.LavaFormatError("critical: failed assigning CU tracker callback, skipping", err,
+			utils.Attribute{Key: "block", Value: block},
+		)
+	} else {
+		k.cuTrackerTS.AddTimerByBlockHeight(ctx, block+blocksToSave-1, []byte(sub.Consumer), []byte(strconv.FormatUint(sub.Block, 10)))
+	}
+}
+
+func (k Keeper) handleZeroDurationLeftForSubscription(ctx sdk.Context, block uint64, blockTime time.Time, sub *types.Subscription) {
+	// subscription duration has already reached zero before and should have
+	// been removed before. Extend duration by another month (without adding
+	// CUs) to allow smooth operation.
+	utils.LavaFormatError("critical: negative duration for subscription, extend by 1 month",
+		fmt.Errorf("negative duration in AdvanceMonth()"),
+		utils.Attribute{Key: "consumer", Value: sub.Consumer},
+		utils.Attribute{Key: "plan", Value: sub.PlanIndex},
+		utils.Attribute{Key: "block", Value: block},
+	)
+	// normally would panic! but can "recover" by auto-extending by 1 month
+	// (don't bother to modify sub.MonthExpiryTime to minimize state changes)
+	expiry := uint64(NextMonth(blockTime).UTC().Unix())
+	k.subsTS.AddTimerByBlockTime(ctx, expiry, []byte(sub.Consumer), []byte{})
+}
+
+func (k Keeper) resetSubscriptionDetailsAndAppendEntry(ctx sdk.Context, sub *types.Subscription, block uint64, blockTime time.Time) {
```

Comment:
the blocktime is always the current block time, why do we need it as an argument?

Reply:
I actually call this function one time with the next epoch time, in the function `upgradeSubscriptionPlan`

Reply:
i mean the blocktime = date

Reply:
You're absolutely right sir!

Fixed in 6b8cea9ceabb6efdfa2d224968ddd45f51f559f5